backend/matchAlgorithm.js

```diff
@@ -1,40 +1,143 @@
+        update: { score: similarity.similarity },
+        create: { adopterId: parseInt(currentUser.id), adopteeId: parseInt(id), score: similarity.similarity }
+      });
+    }
+
+    return userType === 'Adopter'
+      ? { adopter: currentUser, adoptee: adoptees.find(adoptee => adoptee.id === bestMatch) }
+      : { adopter: adopters.find(adopter => adopter.id === bestMatch), adoptee: currentUser };
+  };
+
+  if (userRole === 'Adopter') {
+    const currentAdopter = adopters.find(adopter => adopter.UserId === loggedInUserId);
+    const currentAdopterPrefList = adopterPreferences[currentAdopter.id];
+
+    const rankDifferent = currentAdopterPrefList.length > 1 &&
+      currentAdopterPrefList.some((id, i, arr) => {
+        if (i === 0) return false;
+        return rawAdopterPreferences.find(pref => pref.adopteeId === id && pref.adopterId === currentAdopter.id).rank !==
+          rawAdopterPreferences.find(pref => pref.adopteeId === arr[i - 1] && pref.adopterId === currentAdopter.id).rank;
+      });
+
+    if (rankDifferent) {
+      const freeAdopters = new Set(adopters.map(adopter => adopter.id));
+      const proposerIndex = adopters.reduce((acc, adopter) => ({ ...acc, [adopter.id]: 0 }), {});
+
+      const { userMatches, otherUserMatches } = applyGaleShapley(adopterPreferences, adopteePreferences, freeAdopters, proposerIndex);
+
+      return {
+        adopter: currentAdopter,
+        adoptee: adoptees.find(adoptee => adoptee.id === userMatches[currentAdopter.id])
+      };
+    } else {
+      return await applyCosineSimilarity(currentAdopter, currentAdopterPrefList, 'Adopter');
+    }
   }
-  
-  return matches;
+
+  if (userRole === 'Adoptee') {
+    const currentAdoptee = adoptees.find(adoptee => adoptee.UserId === loggedInUserId);
+    const currentAdopteePrefList = adopteePreferences[currentAdoptee.id];
+
+    const rankDifferent = currentAdopteePrefList.length > 1 &&
+      currentAdopteePrefList.some((id, i, arr) => {
+        if (i === 0) return false;
+        return rawAdopteePreferences.find(pref => pref.adopterId === id && pref.adopteeId === currentAdoptee.id).rank !==
+          rawAdopteePreferences.find(pref => pref.adopterId === arr[i - 1] && pref.adopteeId === currentAdoptee.id).rank;
+      });
+
+    if (rankDifferent) {
+      const freeAdoptees = new Set(adoptees.map(adoptee => adoptee.id));
+      const proposerIndex = adoptees.reduce((acc, adoptee) => ({ ...acc, [adoptee.id]: 0 }), {});
+
+      const { userMatches, otherUserMatches } = applyGaleShapley(adopteePreferences, adopterPreferences, freeAdoptees, proposerIndex);
+
+      return {
+        adopter: adopters.find(adopter => adopter.id === otherUserMatches[currentAdoptee.id]),
+        adoptee: currentAdoptee
+      };
+    } else {
+      return await applyCosineSimilarity(currentAdoptee, currentAdopteePrefList, 'Adoptee');
+    }
+  }
+
+  return { error: "Invalid user role or user not found" };
```

Comment:
This is similar to adopter role. It checks if the preferences have different rank or not and the appropriate matching method is used(Gale-Shapley or cosine similarity)